### DISCUSSION TITLE
no async in sync::watch::reciever::borrow_and_update()

### DISCUSSION THREAD

**Discussion by MagnumTrader** [tokio-rs/tokio#6271] [General] [answered]
Hi,

i have a question about the best implementation.
I have a scenario where a struct updates its state frequently, notifying over a broadcasting channel which the waiters are awaiting in a loop, in turn watchers hold a Arc<RwLock<notifyerStruct>> to access the data and do their thing.

now i stumbled upon the watch channel in the sync module which seems to be what im looking for here! i only care about the latest update since this are financial markets in realtime.

now to my question.

the borrow_and_update() method on reciever is not async.
And according to the docs i should prefer using it over 
`rx.changed().await;
let x = rx.borrow()
// do stuff with x
`
according to [this](https://docs.rs/tokio/latest/tokio/sync/watch/index.html#borrow_and_update-versus-borrow)

but since borrow_and_update() is not async. will this cause problems? 
is there any reason for this to not be async since most functions and methods in tokio are.

best regards
M

**Comment by Darksonn** [comment, marked as answer]
What the docs suggest is using
```rs
rx.changed().await;
let x = rx.borrow_and_update();
```
instead of 
```rs
rx.changed().await;
let x = rx.borrow();
```
So both versions have an `.await`.

**Reply by MagnumTrader** [reply]
oh i think i misunderstood the concept. in my head borrow_and_update() would do both the checking for an update, borrowing and update last seen value but that isnt the intended behavior?


```rust
rx.changed().await;
// additional changes can happen here
let x = rx.borrow();

rx.changed().await;
// additional changes can happen here aswell
let x = rx.borrow_and_update();
```

**Reply by Darksonn** [reply]
The `borrow_and_update` method will never sleep, so it does not need to be async.
